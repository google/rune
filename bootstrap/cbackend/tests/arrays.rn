println "concat array of integers: ",   [1] + [2, 3, 4]
println "concat array of strings: ",    ["a", "b"] + ["c", "d", "e"]
println "concat array of tuples: ",     [(1, "a")] + [(2, "b"), (3, "c"), (4, "d")]
println "concat array of arrays: ",     [[1]] + [[2], [3], [4]]

l = arrayof(u64)
l.append(123)
l.append(456)
println "result of two appends: ", l
println "l.length() = ", l.length()
println "l[0] = ", l[0]
println "l[1] = ", l[1]

l.concat([789])
println "l = ", l
println "l.length() = ", l.length()

l.reverse()
println "l = ", l
println "l[0] = ", l[0]
println "l[1] = ", l[1]
println "l[2] = ", l[2]

// Verify that we can use expressions to index into arrays
a = 1
println "l[a] = ", l[a]
println "l[a+1] = ", l[a + 1]


// arrays of strings
l2 = ["a", "b"]
println "l2 = ", l2
l2.append("c")
println "l2 = ", l2
l2.concat(["d", "e", "f"])
println "l2 = ", l2
l2.reverse()
println "l2 = ", l2
println "l2.length() = ", l2.length()

// arrays of structs
struct Foo {
  name: string
}

l3 = arrayof(Foo)
l3.append(Foo("foo1"))
l3.append(Foo("foo2"))
println "l3 = ", l3
l3.concat([Foo("foo3")])
println "l3 = ", l3
l3.reverse()
println "l3 = ", l3
println "l3.length() = ", l3.length()

// arrays of tuples
l4 = [(1, "a"), (2, "b"), (3, "c")]
l4.append((4, "d"))
println "l4 = ", l4
l4.concat([(5, "e"), (6, "f")])
println "l4 = ", l4
l4.reverse()
println "l4 = ", l4
println "l4.length() = ", l4.length()

// arrays of arrays
l5 = [[(1, "a"), (2, "b")], [(3, "c")]]
println "l5 = ", l5
l5.append([(4, "d"), (5, "e"), (6, "f")])
println "l5 = ", l5
l5.reverse()
println "l5 = ", l5
println "l5.length() = ", l5.length()
