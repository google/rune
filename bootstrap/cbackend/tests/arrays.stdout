concat array of integers: [1, 2, 3, 4]
concat array of strings: ["a", "b", "c", "d", "e"]
concat array of tuples: [(1, "a"), (2, "b"), (3, "c"), (4, "d")]
concat array of arrays: [[1], [2], [3], [4]]
result of two appends: [123, 456]
l.length() = 2
l[0] = 123
l[1] = 456
l = [123, 456, 789]
l.length() = 3
l = [789, 456, 123]
l[0] = 789
l[1] = 456
l[2] = 123
l[a] = 456
l[a+1] = 123
l2 = ["a", "b"]
l2 = ["a", "b", "c"]
l2 = ["a", "b", "c", "d", "e", "f"]
l2 = ["f", "e", "d", "c", "b", "a"]
l2.length() = 6
l3 = [Foo("foo1",), Foo("foo2",)]
l3 = [Foo("foo1",), Foo("foo2",), Foo("foo3",)]
l3 = [Foo("foo3",), Foo("foo2",), Foo("foo1",)]
l3.length() = 3
l4 = [(1, "a"), (2, "b"), (3, "c"), (4, "d")]
l4 = [(1, "a"), (2, "b"), (3, "c"), (4, "d"), (5, "e"), (6, "f")]
l4 = [(6, "f"), (5, "e"), (4, "d"), (3, "c"), (2, "b"), (1, "a")]
l4.length() = 6
l5 = [[(1, "a"), (2, "b")], [(3, "c")]]
l5 = [[(1, "a"), (2, "b")], [(3, "c")], [(4, "d"), (5, "e"), (6, "f")]]
l5 = [[(4, "d"), (5, "e"), (6, "f")], [(3, "c")], [(1, "a"), (2, "b")]]
l5.length() = 3
